pad_server/src/telemetry.c

```diff
@@ -259,26 +262,88 @@ void *telemetry_run(void *arg) {
         header_p hdr = {.type = TYPE_TELEM, .subtype = TELEM_PRESSURE};
         pressure_p body = {.id = 1, .time = time, .pressure = pressure};
 
-        /* Send data to all clients. */
         struct iovec pkt[2] = {
             {.iov_base = &hdr, .iov_len = sizeof(hdr)},
             {.iov_base = &body, .iov_len = sizeof(body)},
         };
         struct msghdr msg = {
-            .msg_name = NULL,
-            .msg_namelen = 0,
             .msg_iov = pkt,
             .msg_iovlen = (sizeof(pkt) / sizeof(struct iovec)),
-            .msg_control = NULL,
-            .msg_controllen = 0,
-            .msg_flags = 0,
         };
         telemetry_publish(&telem, &msg);
-
-        usleep(1000);
+        usleep(1000000);
     }
 
     thread_return(0); // Normal return
 
     pthread_cleanup_pop(1);
+    pthread_cleanup_pop(1);
+}
+
+/*
+ * Helper function to send the entire padstate over telemetry
+ * @param state the pad state
+ * @param sock the telemetry socket
+ */
+void telemetry_send_padstate(padstate_t *state, telemetry_sock_t *sock) {
+    struct timespec time;
+    clock_gettime(CLOCK_MONOTONIC, &time);
+    uint32_t time_ms = time.tv_sec * 1000 + time.tv_nsec / 1000000;
+
+    // sending arming update
+    header_p hdr = {.type = TYPE_TELEM, .subtype = TELEM_ARM};
+    arm_lvl_e arm_lvl = padstate_get_level(state);
+    arm_state_p body = {.time = time_ms, .state = arm_lvl};
+
+    struct iovec pkt[2] = {
+        {.iov_base = &hdr, .iov_len = sizeof(hdr)},
+        {.iov_base = &body, .iov_len = sizeof(body)},
+    };
+    struct msghdr msg = {
+        .msg_iov = pkt,
+        .msg_iovlen = (sizeof(pkt) / sizeof(struct iovec)),
+    };
+    telemetry_publish(sock, &msg);
+
+    // sending actuator update
+    for (int i = 0; i < NUM_ACTUATORS; i++) {
+        header_p hdr = {.type = TYPE_TELEM, .subtype = TELEM_ACT};
+        bool act_state;
+        padstate_get_actstate(state, i, &act_state);
+        act_state_p body = {.time = time_ms, .id = i, .state = act_state};
+
+        struct iovec pkt[2] = {
+            {.iov_base = &hdr, .iov_len = sizeof(hdr)},
+            {.iov_base = &body, .iov_len = sizeof(body)},
+        };
+        struct msghdr msg = {
+            .msg_iov = pkt,
+            .msg_iovlen = (sizeof(pkt) / sizeof(struct iovec)),
+        };
+        telemetry_publish(sock, &msg);
+    }
+}
+
+void *telemetry_update_padstate(void *arg) {
+    telemetry_padstate_args_t *args = (telemetry_padstate_args_t *)arg;
+    padstate_t *state = args->state;
+
+    for (;;) {
+        telemetry_send_padstate(state, args->sock);
+
+        struct timespec cond_timeout;
+        clock_gettime(CLOCK_REALTIME, &cond_timeout);
+        cond_timeout.tv_sec += PADSTATE_UPDATE_TIMEOUT_SEC;
+
+        // waiting until cond_timedwait times out
+        pthread_mutex_lock(&state->update_mut);
+        while (pthread_cond_timedwait(&state->update_cond, &state->update_mut, &cond_timeout) == 0 &&
```

Comment:
This fails to send the update when you get `ETIMEDOUT` from your wait.

I would approach this more like:

```c
int err = -1;
while (err != ETIMEDOUT && !state->update_recorded) {
    /* Loop until update or timeout */
    err = pthread_cond_timedwait(...);
}

/* Once you are here, you know that you've succesfully waited and either gotten an update or timed out waiting */
telemetry_send_padstate(...);
state->update_recorded = false;
```